From e56c9b520e439eee6f2fe46ce7362855fd3a9fce Mon Sep 17 00:00:00 2001
From: Andy Grover <agrover@redhat.com>
Date: Fri, 9 Jan 2015 00:27:30 -0500
Subject: [iser-target] Ignore non TEXT + LOGOUT opcodes for discovery

Message-id: <1420763250-27236-10-git-send-email-agrover@redhat.com>
Patchwork-id: 102565
O-Subject: [RHEL7.1 PATCH e-stor 9/9] iser-target: Ignore non TEXT + LOGOUT opcodes for discovery
Bugzilla: 1058736
RH-Acked-by: Rob Evers <revers@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Nicholas Bellinger <nab@linux-iscsi.org>

This patch adds a check in isert_rx_opcode() to ignore non TEXT + LOGOUT
opcodes when SessionType=Discovery has been negotiated.

Cc: Or Gerlitz <ogerlitz@mellanox.com>
Signed-off-by: Nicholas Bellinger <nab@linux-iscsi.org>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/ulp/isert/ib_isert.c b/drivers/infiniband/ulp/isert/ib_isert.c
index 06fa658..6b6eb8b 100644
--- a/drivers/infiniband/ulp/isert/ib_isert.c
+++ b/drivers/infiniband/ulp/isert/ib_isert.c
@@ -1100,11 +1100,19 @@ isert_rx_opcode(struct isert_conn *isert_conn, struct iser_rx_desc *rx_desc,
 {
  struct iscsi_hdr *hdr = &rx_desc->iscsi_header;
  struct iscsi_conn *conn = isert_conn->conn;
+ struct iscsi_session *sess = conn->sess;
  struct iscsi_cmd *cmd;
  struct isert_cmd *isert_cmd;
  int ret = -EINVAL;
  u8 opcode = (hdr->opcode & ISCSI_OPCODE_MASK);
 
+ if (sess->sess_ops->SessionType &&
+    (!(opcode & ISCSI_OP_TEXT) || !(opcode & ISCSI_OP_LOGOUT))) {
+  pr_err("Got illegal opcode: 0x%02x in SessionType=Discovery,"
+         " ignoring\n", opcode);
+  return 0;
+ }
+
  switch (opcode) {
  case ISCSI_OP_SCSI_CMD:
   cmd = iscsit_allocate_cmd(conn, GFP_KERNEL);
-- 
1.7.1